cudax/include/cuda/experimental/__async/sender/atomic_intrusive_queue.cuh

```diff
@@ -0,0 +1,122 @@
+
+#include <cuda/experimental/__async/sender/prologue.cuh>
+
+namespace cuda::experimental::__async
+{
+template <auto _NextPtr>
+class _CCCL_TYPE_VISIBILITY_DEFAULT __atomic_intrusive_queue;
+
+template <class _Tp, _Tp* _Tp::* _NextPtr>
+class alignas(64) __atomic_intrusive_queue<_NextPtr>
+{
+public:
+  using __node_pointer        = _Tp*;
+  using __atomic_node_pointer = _CUDA_VSTD::atomic<_Tp*>;
+
+  [[nodiscard]]
+  _CUDAX_API auto empty() const noexcept -> bool
+  {
+    return __head_.load(_CUDA_VSTD::memory_order_relaxed) == nullptr;
+  }
+
+  struct _CCCL_TYPE_VISIBILITY_DEFAULT try_push_result
+  {
+    bool __success;
+    bool __was_empty;
+  };
+
+  _CUDAX_API auto try_push_front(__node_pointer __node) noexcept -> try_push_result
+  {
+    _CCCL_ASSERT(__node != nullptr, "Cannot push a null pointer to the queue");
+    __node_pointer __old_head = __head_.load(_CUDA_VSTD::memory_order_relaxed);
+    __node->*_NextPtr         = __old_head;
+    bool __success            = __head_.compare_exchange_strong(__old_head, __node, _CUDA_VSTD::memory_order_acq_rel);
+    if (__success)
+    {
+      __head_.notify_all();
+    }
+    return {__success, __old_head == nullptr};
+  }
+
+  _CUDAX_API auto push_front(__node_pointer __node) noexcept -> bool
+  {
+    _CCCL_ASSERT(__node != nullptr, "Cannot push a null pointer to the queue");
+    __node_pointer __old_head = __head_.load(_CUDA_VSTD::memory_order_relaxed);
+    do
+    {
+      __node->*_NextPtr = __old_head;
+    } while (!__head_.compare_exchange_weak(__old_head, __node, _CUDA_VSTD::memory_order_acq_rel));
+    __head_.notify_all();
+    return __old_head == nullptr;
+  }
+
+  _CUDAX_API void prepend(__intrusive_queue<_NextPtr> __queue) noexcept
+  {
+    if (__queue.empty())
+    {
+      return;
+    }
+
+    __node_pointer __new_head = __queue.front();
+    __node_pointer __tail     = __queue.back();
+    __node_pointer __old_head = __head_.load(_CUDA_VSTD::memory_order_relaxed);
+    __tail->*_NextPtr         = __old_head;
+    while (!__head_.compare_exchange_weak(__old_head, __new_head, _CUDA_VSTD::memory_order_acq_rel))
+    {
+      __tail->*_NextPtr = __old_head;
+    }
+    __queue.clear();
+    __head_.notify_all();
+  }
+
+  _CUDAX_API auto pop_all() noexcept -> __intrusive_queue<_NextPtr>
+  {
+    return __intrusive_queue<_NextPtr>::make(__reset_head());
+  }
+
+  _CUDAX_API auto pop_all_reversed() noexcept -> __intrusive_queue<_NextPtr>
+  {
+    return __intrusive_queue<_NextPtr>::make_reversed(__reset_head());
+  }
+
+  // Wait until the queue has an item in it:
+  _CUDAX_API void wait() noexcept
+  {
+    __head_.wait(nullptr);
+  }
+
+private:
+  _CUDAX_API auto __reset_head() noexcept -> __node_pointer
+  {
+    __node_pointer __old_head = __head_.load(_CUDA_VSTD::memory_order_relaxed);
+    while (!__head_.compare_exchange_weak(__old_head, nullptr, _CUDA_VSTD::memory_order_acq_rel))
+    {
+      ;
+    }
+    return __old_head;
```

Comment:
Why not just use `__head_.exchange(nullptr, _CUDA_VSTD::memory_order_acq_rel)`?

Reply:
I believe this is your code!

Reply:
I don't think I wrote this code.

The original code I wrote was here: https://github.com/facebookexperimental/libunifex/blob/927497ba911bc6fc39e314ba026ba6db05598c93/include/unifex/detail/atomic_intrusive_queue.hpp#L123

Note the use of `atomic::exchange(nullptr)` in the `dequeue_all()` and `dequeue_all_reversed()` functions.

Reply:
looks like it was changed here: https://github.com/NVIDIA/stdexec/commit/38e7a42a965384c3f6f19d751e70c9e921f03ad8

what do you make of that?

Reply:
I don't understand the rationale behind that change in that commit. The semantic result of a cas loop like that is the same as original exchange() call.

Reply:
yeah i think that change was bogus. i wasn't thinking critically when i merged it.